### DISCUSSION TITLE
How to connect to emulators?

### DISCUSSION THREAD

**Discussion by l2aelba** [yume-chan/ya-webadb#245] [Q&A] [answered]
I was wondering about does it possible to connect to emulators like Bluestack / Nox / LD Player or something?

Or any trick for it? like by installed some program or something to fake it.

Thanks

**Comment by yume-chan** [comment]
I don't know about those third-party emulators, but at least for the official [Android Emulator](https://developer.android.com/studio/run/emulator), it's possible, with some help of websockify.

![image](https://user-images.githubusercontent.com/1330321/105126852-9579d900-5b1a-11eb-97f5-d3caf8275a99.png)

If the ones you have mentioned can also be automatically picked up by a normal `adb`, then this method will also apply to them.

It's a little complex and requires some programming skill, please tell me if you want to know the details, then I will find some spare time to organize the steps and post here.

**Comment by yume-chan** [comment, marked as answer]
I added detection for WebSocket backend in ae1f9a223ecb8953eba24e33187b93dd9afc9166 to make it easier.

----

Some background information:

1. `adb` connects to AVD (Android Virtual Device, the official emulator) using TCP sockets.
2. AVD may choose any odd port number between `5555` and `5587` (even numbers are for serial connection), `adb devices` will scan all ports in that range to detect emulators.

So you need to firstly find out which port your emulator is currently using.

Run `adb devices`, emulators have serial numbers in form of "emulator-55XX" (where `XX` is two digits), add `1` to that number, that's the emulator's ADB port.

```text
PS E:\AndroidSDK\platform-tools> ./adb devices
List of devices attached
emulator-5554   device
```

Or you may just try `5555`, normally emulator starts at this port.

----

1. Web browsers can't open TCP sockets.
2. Web browsers can use WebSocket, which is similar to TCP socket, so a native program can convert between TCP socket and WebSocket to bridge the gap.

I use websockify-js here, other similar programs should also work.

1. Install [Node.js](https://nodejs.org/en/)
2. Clone (if you have Git) or download the source code of [novnc/websockify-js](https://github.com/novnc/websockify-js) repository
3. In a terminal, navigate to the `websockify` folder, run `npm i` (this command only needs to be run once)

Every time you want to connect to your emulators, in the `websockify` folder, run `node websockify.js 15555 localhost:5555`, replace `5555` to the port you found previously, but DO NOT change `15555`. The output should be like

```text
PS E:\git\novnc\websockify-js\websockify> node .\websockify.js 15555 localhost:5555
WebSocket settings:
    - proxying from :15555 to localhost:5555
    - Running in unencrypted HTTP (ws://) mode
```

Open my [demo](https://yume-chan.github.io/ya-webadb) in browser or run it yourself, wait a few seconds, you should see `ws://localhost:15555` in available devices

![image](https://user-images.githubusercontent.com/1330321/108490317-99e20f00-72dd-11eb-817b-b4a051add130.png)

Click <kbd>Connect</kbd>, the confirmation dialog should display on your emulator screen.

In my testing the confirmation dialog sometimes [closes by its own after a brief period](https://stackoverflow.com/questions/63922631/android-emulator-allow-usb-debugging-dialog-close-right-after-showing). The only solution I found was simply click <kbd>Allow</kbd> fast enough.

----

Bouns:

`adb connect` command (ADB over Wi-Fi) also uses TCP sockets to connect to devices wirelessly. So this method also applies to ADB over Wi-Fi.

**Reply by l2aelba** [reply]
Thanks, I will take a look

**Reply by guanzhangrtk** [reply]
@yume-chan Thanks for this writeup! I followed the instructions and was able to connect to a device wirelessly, however the only thing that doesn't work is `scrcpy`: it popped up a dialog saying: `Error: more than one device/emulator`, any ideas why? Screen Capture works fine though, so I'm a bit baffled -- thanks!

**Reply by yume-chan** [reply]
@guanzhangrtk Are you using the native scrcpy client or my web clone?

**Reply by guanzhangrtk** [reply]
@yume-chan Your web clone:
![image](https://user-images.githubusercontent.com/44019025/123935520-455b4880-d9cf-11eb-895d-e2b004acb270.png)
Everything works except for `Scrcpy` -- for instance Screen Capture works:
![image](https://user-images.githubusercontent.com/44019025/123935701-6b80e880-d9cf-11eb-8bcf-b63792f22af4.png)
I do also use the native `scrcpy` client but not at the same time, thanks so much!

**Reply by guanzhangrtk** [reply]
So the only thing I can think of that I'm doing differently is that I'm using node v16.2.0 instead of the recommended versions but I didn't see any errors associated with this change. Other than that I also tried updating `scrcpy-server` to v1.18 but had the same issue. How do we go about getting more debugging info? Thanks!

EDIT: Actually another fundamental difference is that the Android Device is not connected via localhost but via an IP address on the LAN

**Reply by guanzhangrtk** [reply]
![image](https://user-images.githubusercontent.com/44019025/123980166-49518f80-d9fc-11eb-98f5-70ded5bbacf8.png)
Okay finally got it working, turns out I need to enable the "Use forward connections" option and then it worked. I think the error message is misleading since the actual error is truncated (on the second line), I think the full error message should be:

```
adb: error: more than one device/emulator
ERROR: "adb reverse" returned with value 1
WARN: 'adb reverse' failed, fallback to 'adb forward'
```
as mentioned in `scrcpy`'s [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md#several-devices-connected).

Now the question is why do I need to enable this option in the first place. The Android device is running Android 7.1.1 and also when I use the native `scrcpy` binary I didn't have this issue. As a side note when I tried to use v1.18 and enabled that option, it gave an empty Error dialog box (no message).

@yume-chan is this something you're willing to look into? I see great potential in this demo and really hope that it can become a full-fledged project. I'm definitely going to try to share it with the Android community to get more people interested in it. Please let me know your thoughts.

I do have a couple more bugs/issues that I've found but will hold off on posting them until I see your response. Thanks again for making this project available!

**Reply by yume-chan** [reply]
> I think the full error message should be:
>
> ```
> adb: error: more than one device/emulator
> ERROR: "adb reverse" returned with value 1
> WARN: 'adb reverse' failed, fallback to 'adb forward'
> ```

I still didn't get you. This error message is from native ADB server, but since you are using websockify that connects to your device directly, there is no ADB server (the part on your device is called ADB daemon, ADB client/server are both the `adb` binary running on your computer). So I have no idea where you got this.

> Now the question is why do I need to enable this option in the first place.

That FAQ already explained the reason. ADB daemon has a bug that reverse tunnel doesn't work when connected wirelessly. Scrcpy knows you are using ADB over WiFi so it switches to forward tunnel automatically. My Web app can't detect that (WebSocket backend doesn't equal to ADB over WiFi), so it only has an option to let users manually opt-in.

Anyway, using the WebSocket backend is not a supported scenario, bug fixes for that will be on low priority.

> As a side note when I tried to use v1.18 and enabled that option, it gave an empty Error dialog box (no message).

I haven't tried 1.18. Scrcpy releases usually have incompatible startup arguments, so using different versions of clients and servers won't work at most time. I'll check that later.

> I see great potential in this demo and really hope that it can become a full-fledged project.

It's a side project, for learning, for fun. I don't have such a high expectation. Of course, contributions are welcomed, including bug reports, feature requests, and PRs.

**Reply by guanzhangrtk** [reply]
> I still didn't get you. This error message is from native ADB server, but since you are using websockify that connects to your device directly, there is no ADB server (the part on your device is called ADB daemon, ADB client/server are both the `adb` binary running on your computer). So I have no idea where you got this.

From your GUI, the error message was simply "more than one device/emulator". The three line error message I posted above I copied and pasted from scrcpy's FAQ to illustrate that potentially the message was truncated in your GUI. So are you saying in your implementation of adb it just has that one line error message?

In any case that was the only issue I had, everything else worked perfectly and is good enough for demonstration purposes. The only suggestion I have is perhaps having the GUI remember that setting (at least for that session) such that if I switch to the shell mode and back I don't have to re-enable that feature again, but those are minor stuff.

tinyh264 does use up quite a bit of CPU cycles as well as RAM, but not sure if there are better alternatives.

Thanks for the response.

**Reply by yume-chan** [reply]
> [...] potentially the message was truncated in your GUI

Those two lines are from scrcpy. ADB daemon only returns the first line.

This issue had been fixed in Android 9, so I don't have a device to reproduce and fix it. If you have any methods to reliably detect Android < 9 or this error message when initiating the scrcpy module, and automatically switch to forward connection, PRs are welcomed.

The upstream issue is https://issuetracker.google.com/issues/37066218

> The only suggestion I have is perhaps having the GUI remember that setting (at least for that session) such that if I switch to the shell mode and back I don't have to re-enable that feature again

Maybe I will add that when I have time.

**Reply by pr4x77** [reply]
I cant get this work. Getting the following error:

Error:
```
Error: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.
    at w.connect (https://yume-chan.github.io/ya-webadb/_next/static/chunks/pages/_app-7d11a706073519ba.js:1:1700)
    at Object.onClick (https://yume-chan.github.io/ya-webadb/_next/static/chunks/pages/_app-7d11a706073519ba.js:1:4708)
    at r._onClick (https://yume-chan.github.io/ya-webadb/_next/static/chunks/pages/_app-7d11a706073519ba.js:1:126599)
    at Object.eU (https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:14912)
    at eH (https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:15066)
    at https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:33374
    at re (https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:33473)
    at rn (https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:33887)
    at https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:39345
    at oP (https://yume-chan.github.io/ya-webadb/_next/static/chunks/framework-e1850a8151a6a66f.js:9:96345)
```

If I use encrypted websockets I get the following error:

`Error: WebSocket connect failed
    at e.onerror (https://yume-chan.github.io/ya-webadb/_next/static/chunks/pages/_app-7d11a706073519ba.js:1:1808)`

Any solution for this?

**Reply by yume-chan** [reply]
The error message is very clear. Either run WebSockify on the same device where you run Tango, or add a valid HTTPS certificate to the WebSockify server.

**Reply by 1234567890nevver** [reply]
devices attached
emulator-5554   device
Or you may just try 5555,

**Reply by Ppeepost4489** [reply]
> I added detection for WebSocket backend in ae1f9a223ecb8953eba24e33187b93dd9afc9166 to make it easier.
> 
> ----
> 
> Some background information:
> 
> 1. `adb` connects to AVD (Android Virtual Device, the official emulator) using TCP sockets.
> 2. AVD may choose any odd port number between `5555` and `5587` (even numbers are for serial connection), `adb devices` will scan all ports in that range to detect emulators.
> 
> So you need to firstly find out which port your emulator is currently using.
> 
> Run `adb devices`, emulators have serial numbers in form of "emulator-55XX" (where `XX` is two digits), add `1` to that number, that's the emulator's ADB port.
> 
> ```text
> PS E:\AndroidSDK\platform-tools> ./adb devices
> List of devices attached
> emulator-5554   device
> ```
> 
> Or you may just try `5555`, normally emulator starts at this port.
> 
> ----
> 
> 1. Web browsers can't open TCP sockets.
> 2. Web browsers can use WebSocket, which is similar to TCP socket, so a native program can convert between TCP socket and WebSocket to bridge the gap.
> 
> I use websockify-js here, other similar programs should also work.
> 
> 1. Install [Node.js](https://nodejs.org/en/)
> 2. Clone (if you have Git) or download the source code of [novnc/websockify-js](https://github.com/novnc/websockify-js) repository
> 3. In a terminal, navigate to the `websockify` folder, run `npm i` (this command only needs to be run once)
> 
> Every time you want to connect to your emulators, in the `websockify` folder, run `node websockify.js 15555 localhost:5555`, replace `5555` to the port you found previously, but DO NOT change `15555`. The output should be like
> 
> ```text
> PS E:\git\novnc\websockify-js\websockify> node .\websockify.js 15555 localhost:5555
> WebSocket settings:
>     - proxying from :15555 to localhost:5555
>     - Running in unencrypted HTTP (ws://) mode
> ```
> 
> Open my [demo](https://yume-chan.github.io/ya-webadb) in browser or run it yourself, wait a few seconds, you should see `ws://localhost:15555` in available devices
> 
> ![image](https://user-images.githubusercontent.com/1330321/108490317-99e20f00-72dd-11eb-817b-b4a051add130.png)
> 
> Click <kbd>Connect</kbd>, the confirmation dialog should display on your emulator screen.
> 
> In my testing the confirmation dialog sometimes [closes by its own after a brief period](https://stackoverflow.com/questions/63922631/android-emulator-allow-usb-debugging-dialog-close-right-after-showing). The only solution I found was simply click <kbd>Allow</kbd> fast enough.
> 
> ----
> 
> Bouns:
> 
> `adb connect` command (ADB over Wi-Fi) also uses TCP sockets to connect to devices wirelessly. So this method also applies to ADB over Wi-Fi. 

pm grant com.phascinate.precisevolume android.permission.DUMP
adb shell pm grant com.phascinate.precisevolume android.permission.DUMP

**Reply by Ppeepost4489** [reply]
Run

**Reply by evgenvl1285-spec** [reply]
PS E:\AndroidSDK\platform-tools> ./adb devices
List of devices attached
emulator-5554   device

**Comment by yume-chan** [comment]
Now it's possible to connect to emulators and ADB over WiFi without webscokify. See #349.

**Comment by ubipo** [comment]
Is it possible to connect to an adb server (e.g. started with `adb -a -P 5037 nodaemon server start`) with this technique? I get a 'Connection closed unexpectedly' error. The adb server outputs the following:
```
adb D 04-16 19:05:07 111186 111186 fdevent_epoll.cpp:174] (fdevent 2: fd 10 R) got events 0x1
adb D 04-16 19:05:07 111186 111186 sockets.cpp:457] LS(3): created (fd=11)
adb D 04-16 19:05:07 111186 111186 sockets.cpp:973] Connecting to smart socket
adb D 04-16 19:05:07 111186 111186 sockets.cpp:961] Creating smart socket
adb D 04-16 19:05:07 111186 111186 sockets.cpp:968] SS(0)
adb D 04-16 19:05:07 111186 111186 fdevent_epoll.cpp:174] (fdevent 8: fd 11 R) got events 0x1
adb D 04-16 19:05:07 111186 111186 sockets.cpp:386] LS(3): event_func(fd=11(==11), ev=0001)
adb D 04-16 19:05:07 111186 111186 sockets.cpp:187] LS(3): post adb_read(fd=11,...) r=24 (errno=0) avail=1048576
adb D 04-16 19:05:07 111186 111186 sockets.cpp:187] LS(3): post adb_read(fd=11,...) r=-1 (errno=11) avail=1048552
adb D 04-16 19:05:07 111186 111186 sockets.cpp:203] LS(3): fd=11 post avail loop. r=-1 is_eof=0 forced_eof=0
adb D 04-16 19:05:07 111186 111186 sockets.cpp:792] SS(0): enqueue 24
adb D 04-16 19:05:07 111186 111186 sockets.cpp:808] SS(0): bad size (4294967295)
adb D 04-16 19:05:07 111186 111186 sockets.cpp:951] SS(0): closed
adb D 04-16 19:05:07 111186 111186 sockets.cpp:347] entered local_socket_close. LS(3) fd=11
adb D 04-16 19:05:07 111186 111186 sockets.cpp:333] LS(3): destroying fde.fd=11
adb D 04-16 19:05:07 111186 111186 sockets.cpp:369] LS(3): closed
adb D 04-16 19:05:07 111186 111186 sockets.cpp:219] LS(3): fd=11 post peer->enqueue(). r=-1
adb D 04-16 19:05:07 111186 111186 fdevent_epoll.cpp:174] (fdevent 9: fd 11 R) got events 0x5
```

For context: connecting via USB does work and I tried websockify-js, the original Python websockify, and websocat, but none of them seem to work.

**Reply by ubipo** [reply]
Ah huh I might be confusing adb over usb and the communication with the adb daemon server (https://gist.github.com/hfutxqd/a5b2969c485dabd512e543768a35a046). I might look into implementing the latter if those really are separate things.

**Reply by yume-chan** [reply]
No, Tango implements the protocol between ADB server and ADB daemon. I believe it's possible to create a "fake" backend to parse and re-package messages to be sent to ADB server.

First it should query the ADB server for connected devices. Because ADB server has already authenticated with the devices, this backend doesn't need `Adb.authenticate()`, but should use `new Adb()` directly.

Then it acts as a backend, but instead of forwarding packages blindly, it needs to parse those packages and translate them to the protocol the other end can understand.

It may itself accepts another backend for communicating with ADB server (`net` module in Node.js or Direct Sockets API in browsers).

The forward and reverse tunnels are extra tricky:

Tango doesn't need forward tunnel registration, the users can connect to forward tunnels using `adb.createSocket("tcp:8080")`. This package needs to call `adb forward tcp:0 tcp:8080`, connect to the returned port, then remove the tunnel.

Same for reverse tunnel, Tango allows any string for local address, but this package needs to choose a port, listen on it, call `adb reverse tcp:1234 tcp:8080`, and generate fake ADB sockets for each incoming connections.

**Reply by yume-chan** [reply]
Now Tango supports the protocol between ADB client and ADB server: https://tango-adb.github.io/docs/tango/server/connection. I only have a server connection that based on Node.js `net` module. A WebSocket server connection class (similar to WebSocket daemon connection) is required if you want to use WebSocket.

**Comment by iamqiqi1017** [comment]
Why not integrate the method of connecting AVD into your project instead of installing another project separately, which is easier to deploy and apply.
